### PR TITLE
Update KvbAppFilter and TRS to Handle Event Groups

### PR DESCRIPTION
**Pull request by shgandhi**
This PR updates KvbAppFilter and TRS to handle event groups. Please see individual commits for more detail.
Note - This PR does not include the corresponding unit tests  yet.

### CODE REVIEW

**Review by robem — APPROVED** [review]
Thanks for working on this 🚀  and the commit messages 💪  Looks good to me. Some nits. There are some more `const` references possible that I didn't flag.
Did you do any manual runtime tests or is that not possible yet?

File: kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp [hunk 1]
```diff
@@ -154,24 +188,60 @@ void KvbAppFilter::readBlockRange(BlockId block_id_start,
...
+  LOG_DEBUG(logger_, "readEventGroupRange event_group_id: " << event_group_id << " to " << event_group_id_end);
+
+  for (; event_group_id <= event_group_id_end; ++event_group_id) {
+    std::string cid;
+    auto event_group = getEventGroup(event_group_id, cid);
+    if (event_group.events.empty()) {
+      std::stringstream msg;
+      msg << "Couldn't retrieve event group for event_group_id: " << event_group_id;
+      throw KvbReadError(msg.str());
+    }
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp]: Thanks for catching, updated the message to reflect that.

File: thin-replica-server/include/thin-replica-server/thin_replica_impl.hpp [hunk 2]
```diff
@@ -210,32 +210,50 @@ class ThinReplicaImpl {
...
-      config_->subscriber_list.removeBuffer(live_updates);
-      live_updates->removeAllUpdates();
-      metrics_.subscriber_list_size.Get().Set(config_->subscriber_list.Size());
-      metrics_.updateAggregator();
-      return grpc::Status(grpc::StatusCode::CANCELLED, error.what());
-    } catch (std::exception& error) {
-      LOG_ERROR(logger_, error.what());
-      config_->subscriber_list.removeBuffer(live_updates);
-      live_updates->removeAllUpdates();
+      bool event_group_enabled = true;
```
[the remark this answers is not in the document]
Comment (shgandhi) [reply thin-replica-server/include/thin-replica-server/thin_replica_impl.hpp]: Removed the flag.

**Comment by shgandhi** [comment]
> Thanks for working on this 🚀 and the commit messages 💪 Looks good to me. Some nits. There are some more `const` references possible that I didn't flag.
> Did you do any manual runtime tests or is that not possible yet?

Thanks for the review @robem. I have updated the PR acc. to your comments and  added the TRS changes as well to this PR, so would require your review again. I will now add the unit tests and update the PR acc. if required.

**Review by robem — APPROVED** [review]
I see you chose to duplicate the existing code and adapt it for EventGroups. The change itself is clear to me and I'm ok with this solution but we should do some refactoring in the near future otherwise this might become a maintenance nightmare. Well, the *_impl.hpp wasn't too pretty to begin with :)

💪  Great job so far and thanks for the hard work!

**Comment by shgandhi** [comment]
Thanks for the review @robem. I agree we will need to refactor the code soon.
I added another commit that adds missing header guards in db_categories.h.

**Pull request merged**